balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,626 @@
+			if b.addressIndex.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass(state.ConnectionError)
+		case connectivity.Idle:
+			// A subconn can transition from CONNECTING directly to IDLE when
+			// a transport is successfully created, but the connection fails before
+			// the subconn can send the notification for READY. We treat this
+			// as a successful connection and transition to IDLE.
+			curAddr := b.addressIndex.currentAddress()
+			if activeSD, found := b.subConns.Get(curAddr); !found || activeSD != sd {
+				return
+			}
+			b.state = connectivity.Idle
+			b.addressIndex.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.Idle,
+				Picker:            &idlePicker{exitIdle: b.ExitIdle},
+			})
+		}
+		return
+	}
+
+	// We have finished the first pass, keep re-connecting failing subconns.
+	switch state.ConnectivityState {
+	case connectivity.TransientFailure:
+		sd.lastErr = state.ConnectionError
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: state.ConnectionError},
+		})
+		// We don't need to request re-resolution since the subconn already does
+		// that before reporting TRANSIENT_FAILURE.
+	case connectivity.Idle:
+		sd.subConn.Connect()
+	}
+}
+
+func (b *pickfirstBalancer) endFirstPass(lastErr error) {
+	b.firstPass = false
+	b.state = connectivity.TransientFailure
+
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: connectivity.TransientFailure,
+		Picker:            &picker{err: lastErr},
+	})
+	// Start re-connecting all the subconns that are already in IDLE.
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.state == connectivity.Idle {
+			sd.subConn.Connect()
+		}
+	}
+}
+
+type picker struct {
+	result balancer.PickResult
+	err    error
+}
+
+func (p *picker) Pick(balancer.PickInfo) (balancer.PickResult, error) {
+	return p.result, p.err
+}
+
+// idlePicker is used when the SubConn is IDLE and kicks the SubConn into
+// CONNECTING when Pick is called.
+type idlePicker struct {
+	exitIdle func()
+}
+
+func (i *idlePicker) Pick(balancer.PickInfo) (balancer.PickResult, error) {
+	i.exitIdle()
+	return balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+}
+
+// addressList manages sequentially iterating over addresses present in a list of
+// endpoints. It provides a 1 dimensional view of the addresses present in the
+// endpoints.
+// This type is not safe for concurrent access.
+type addressList struct {
+	addresses []resolver.Address
+	idx       int
+}
+
+func (al *addressList) isValid() bool {
+	return al.idx < len(al.addresses)
+}
+
+func (al *addressList) size() int {
+	return len(al.addresses)
+}
+
+// increment moves to the next index in the address list. If at the last address
+// in the address list, moves to the next endpoint in the endpoint list.
+// This method returns false if it went off the list, true otherwise.
```

Comment:
This part of the comment is no longer valid, right?

// If at the last address
// in the address list, moves to the next endpoint in the endpoint list.
// This method returns false if it went off the list, true otherwise.